encryption.go

```diff
@@ -0,0 +1,413 @@
+			message_chunks = append(message_chunks, message[i:i+e.SIXTEN_KB])
+		}
+		for _, message_chunk := range message_chunks {
+			encrypted_chunk := e.encryptMessage(message_chunk, host)
+			encrypted_message = append(encrypted_message, encrypted_chunk...)
+		}
+	} else {
+		content_type = "multipart/encrypted"
+		encrypted_message = e.encryptMessage(message, host)
+	}
+
+	encrypted_message = append(encrypted_message, e.MIME_BOUNDARY...)
+	encrypted_message = append(encrypted_message, []byte("--\r\n")...)
+
+	req, err := http.NewRequest("POST", endpoint, bytes.NewBuffer(encrypted_message))
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Set("User-Agent", "Bob WinRM client")
+	req.Header.Set("Connection", "Keep-Alive")
+	req.Header.Set("Content-Length", fmt.Sprintf("%d", len(encrypted_message)))
+	req.Header.Set("Content-Type", content_type+";protocol=\""+string(e.protocolString)+"\";boundary=\"Encrypted Boundary\"")
+
+	resp, err := e.ntlmhttp.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("unknown error %w", err)
+	}
+
+	body, err := e.ParseEncryptedResponse(resp)
+
+	return string(body), err
+}
+
+/*
+Takes in the encrypted response from the server and decrypts it
+
+:param response: The response that needs to be decrytped
+:return: The unencrypted message from the server
+*/
+func (e *Encryption) ParseEncryptedResponse(response *http.Response) ([]byte, error) {
+	contentType := response.Header.Get("Content-Type")
+	if strings.Contains(contentType, fmt.Sprintf(`protocol="%s"`, e.protocolString)) {
+		return e.decryptResponse(response, response.Request.URL.Hostname())
+	}
+	body, err := ioutil.ReadAll(response.Body)
+	response.Body.Close()
+	if err != nil {
+		return nil, err
+	}
+	return body, nil
+}
+
+func (e *Encryption) encryptMessage(message []byte, host string) []byte {
+	messageLength := []byte(fmt.Sprintf("%d", len(message)))
+	encryptedStream, _ := e.buildMessage(message, host)
+
+	messagePayload := bytes.Join([][]byte{
+		e.MIME_BOUNDARY,
+		[]byte("\r\n"),
+		[]byte("\tContent-Type: " + string(e.protocolString) + "\r\n"),
+		[]byte("\tOriginalContent: type=application/soap+xml;charset=UTF-8;Length=" + string(messageLength) + "\r\n"),
+		e.MIME_BOUNDARY,
+		[]byte("\r\n"),
+		[]byte("\tContent-Type: application/octet-stream\r\n"),
+		encryptedStream,
+	}, []byte{})
+
+	return messagePayload
+}
+
+func deleteEmpty(b [][]byte) [][]byte {
+	var r [][]byte
+	for _, by := range b {
+		if len(by) != 0 {
+			r = append(r, by)
+		}
+	}
+	return r
+}
+
+func (e *Encryption) decryptResponse(response *http.Response, host string) ([]byte, error) {
+	body, _ := ioutil.ReadAll(response.Body)
+	parts := deleteEmpty(bytes.Split(body, []byte(MimeBoundary+"\r\n")))
+	var message []byte
+
+	for i := 0; i < len(parts); i += 2 {
+		header := parts[i]
+		payload := parts[i+1]
+
+		expectedLengthStr := bytes.SplitAfter(header, []byte("Length="))[1]
+		expectedLength, _ := strconv.Atoi(string(bytes.TrimSpace(expectedLengthStr)))

```

Comment:
I think there the error shouldn't be thrown away, we could be fed with invalid content and decryption should be stopped at this point and an error raised.

Reply:
The parsing seems a bit fragile here. It seems to be encoded as a regular mime message, so I think it would make sense to use a mime parsing library to extract properly the parts (maybe [this](https://pkg.go.dev/mime/multipart) would be enough?)

Reply:
checking error, attempted to use mime/multipart, but it failed - added comment for same on method